### DISCUSSION TITLE
How to properly use React Server Components w/React 18 and Next12?

### DISCUSSION THREAD

**Discussion by milanbgd011** [vercel/next.js#30854] [App Router]
There are only 2 examples of using Next12 and React18 for React Server Components, that is it. Many details are not precisely defined and I spent a bit of time trying to figure out and clarify strict rules for its usage, but info is scarce.

**Questions are:**
1. inside `/pages` directory, do we only use "index.js" as starting point or we can use "index.server.js" and "index.client.js"? what is best practice here and why? are there any pitholes?
2. it seems like all `*.server.js` and `*.client.js` components must be outside the `/pages` directory for it to work properly? if that is not true, we need strong confirmation that it is fine.
3. can we mix server and client components multiple times (load .server.js which loads .client.js, which loads .server.js...)? this should work no problems, right?
4. why `<Head />` is not allowed inside `*.server.js` and `*.client.js` components? how to set page title if we must use `*.server.js` in pages directory
5. is this anti-pattern or good usage on this RSC example (we loose `<Head />` here)? https://github.com/vercel/next-server-components/blob/master/pages/index.js
6. if you create `/pages/index.js` and in there use ie. `components/SomeComponent.server.js` it will be visible in "view-source" as rendered, but will remain BLANK on the site (like hydration removes it). When you change to `/pages/index.server.js` then all works as intended? why?
7. what is recommended way to do "fetch" in `*.server.js` and `*.client.js`? is "hackernews" example the best example and we should stick to this? can we use "SWR" as best solution for fetching here?

If you know the answer, please share.

If you have any more weird and undocumented stuff you want to know, add more questions here.

I suppose @leerob could provide some depth and create maybe best practice how-to. Hackernews example is more demonstrating differences, the Notes app is too complex to be good as example. It would be perfect if we could have a video where all these important details are defined into "recommended" approach. Inner workings are not clear, some stuff should not be done, some should. I guess this would be solved if we get all above questions answered and then new video or repo "best practice" with all variations (using *.js, *.server.js and *.client.js in /pages dir and all details starting from those three points).

**Comment by milanbgd011** [comment]
Another one. When we switch "concurrentFeatures" on, it seems that "process.env" is no longer available in "getServerSideProps" and neither is Axios there. Entire app falls apart when this switch is ON. How will this transition be handled, what will happen with getServerSideProps once we enable concurrentFeatures when all reaches stable state?

**Comment by bbigras** [comment]
> it seems like all *.server.js and *.client.js components must be outside the /pages directory for it to work properly? if that is not true, we need strong confirmation that it is fine.

The demo https://github.com/vercel/next-rsc-demo/tree/f5fc94da5ecd52f3c3ffac482cf7404f98a08e70/pages has some `.server.js` in `pages`.

But I didn't figure out how to use `index.server.tsx`.

**Reply by milanbgd011** [reply]
I saw that, I analyzed every single bit of info there is, but explanation if something is allowed or not is not clear.

I noticed, as described above, that some things do not work and some stuff behaves weird.

That is why we need some ground rules, how the system is build and configured, so we can test it.

**Reply by milanbgd011** [reply]
Based on info we get here, I can test to see if the system works if being used like it was designed.
If anything does not work, I can report issue and reproduction, so it can be looked into.

**Comment by leerob** [comment]
Hey! Thank you for providing all of this info. Server Components are still in alpha and we have a lot of work to do around documentation and education 😄 As it progresses out of alpha and into beta, we'll be adding more documentation and guidance on how to try out these early features.

**Reply by milanbgd011** [reply]
Thank you @leerob very much, really appreciate your work. If it is possible to provide extremely short answers to the questions I asked at the top, it would mean a lot.

For me to set all up properly and try to work with server components and find out about bugs and problems so I can create real issues with repro code attached. This way I know I set all up the way your team imagined. Those Qs are really precise I would say, really short answers needed.

I know it is Alpha, but it does not mean anything if we do not have some basic rules. The internals of this implementation is really complex to figure on my own and for most of devs out there.
